### ISSUE TITLE
Locate missing/extra braces for parsing error

### ISSUE THREAD

**Issue by eight04** [eslint/eslint#7703]
**Tell us about your environment**

* **ESLint Version:** 3.9.1
* **Node Version:** 7.0.0
* **npm Version:** 3.10.8

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

```
{
    "rules": {
        "curly": "error"
    }
}
```

**What did you do? Please include the actual source code causing the issue.**

```js
function test(a) {
    if (a) 
        doSomething();
    }
    
    doSomethingElse();
}

test(a);
```

**What did you expect to happen?**
```
  2:5  error  Expected { after 'if' condition  curly
  7:1  error  Parsing error: Unexpected token }

✖ 2 problem (2 error, 0 warnings)
```

**What actually happened? Please include the actual, raw output from ESLint.**
```
  7:1  error  Parsing error: Unexpected token }

✖ 1 problem (1 error, 0 warnings)
```

Actually, I think it works as intended since ESLint have to build the AST first. But it would be nice if we can get the line number of missing braces. The advantage becomes bigger while working on large files.

Perhaps to make the parser generate a "partial AST" for specific parsing error, or let the rule be able to do some additional check that doesn't depend on the AST.

**Comment by eight04** [comment]
Additionally, it would be possible to insert `{` automatically within `--fix` option.

**Comment by platinumazure** [comment]
In terms of finding a line where an opening brace should be inserted, how would you solve this case?

```js
if (foo)
    if (bar)
        something();
}
```

Which if statement should have the brace? You could put it on either line and it would be syntactically correct. I don't think this is a problem we could solve easily in the parser or the linter.

**Comment by not-an-aardvark** [comment]
I'm not sure this is feasible, because there are several possible places where the missing brace could be inserted, and all of them would mean something different. ESLint wouldn't be able to figure out which place the user meant to put the `{`.

The syntax would be valid with the brace inserted here:

```diff
function test(a) {
-    if (a) 
+    if (a) {
        doSomething();
    }
    
    doSomethingElse();
}

test(a);
```

Or here:

```diff
function test(a) {
    if (a) 
        doSomething();
+   {
    }
    
    doSomethingElse();
}

test(a);
```

Or here:


```diff
function test(a) {
    if (a) 
        doSomething();
    }
+{
    doSomethingElse();
}

test(a);
```

Or here:

```diff
function test(a) {
    if (a) 
        doSomething();
    }
    doSomethingElse();
+{
}

test(a);
```

When a file has invalid syntax, there are a lot of things that the user *might* have meant to do, but it's very difficult for the parser to figure out which one they meant. The best option is for the parser to just report an error at the first place it knows something went wrong, which is what it currently does. I don't think it would be a good idea for ESLint to pick a location and fix it, because it could easily end up with code that the user didn't expect.

**Comment by eight04** [comment]
The main issue is not to guess the position of the braces, but that we should be able to verify some rules (curly, indent, etc.) before the AST is completely built.

Anyway, here is the possible output *with `curly` rule*:
```
if (foo)
    if (bar)
        something();
}
```
Without `--fix`:
```
1:9  error  Expected { after 'if' condition  curly
2:13  error  Expected { after 'if' condition  curly
4:1  error  Parsing error: Unexpected token }
```
With `--fix`, generate output as linting following code:
```
if (foo) {
    if (bar) {
        something();
}
```
```
5:1  error  Parsing error: Unexpected token
```

**Comment by platinumazure** [comment]
Unfortunately, that's a consequence of the architecture.

From http://eslint.org/docs/about/:

> ESLint is designed to have all rules completely pluggable. The default rules are written just like any plugin rules would be. They can all follow the same pattern, both for the rules themselves as well as tests.

We intentionally want to treat all rules the same, and many rules need the full AST to function.

Additionally, from our [README](https://github.com/eslint/eslint/blob/master/README.md#how-does-eslint-performance-compare-to-jshint):

> This is because ESLint uses Espree to construct an AST before it can evaluate your code whereas JSHint evaluates your code as it's being parsed.

So your question essentially becomes: Wouldn't it be great to run rules in two passes, one pre-AST and one post-AST, and be able to flag rules as not needing an AST and being able to run with regular expressions only? And the answer is, sure, I suppose, but I don't see a compelling need for it (and it would need to be a *very* compelling need) to justify the massive overhaul of the architecture (and rewriting of those rules) which would be required. I just don't see it happening.

Parse errors should be a one-time thing as you're linting your codebase for the first time. Then, using continuous integration and (of course) using ESLint, you would never see them again.

**Comment by eight04** [comment]
No need to invest other mechanic like Regex, just lint the valid part before the parsing error.
```
function test(a) {
    if (a) 
        doSomething();
    }
    
    doSomethingElse();
}

test(a);
```
Error at line 7, than let's just lint lines 1~6
```
function test(a) {
    if (a) 
        doSomething();
    }
    
    doSomethingElse();
```
Not all parsing error would be able to determine such "valid part" (also it should be determine by the parser instead of line number), but it works for missing braces since it is always possible to find a valid sequence while balancing the braces unless the first token is `}`.

Still have to mark some rules to `requireFullAST` though.

**Comment by platinumazure** [comment]
Again, looking at this from a cost-benefit analysis, I don't think the benefits are anywhere near the massive costs (rewriting the parser and ESLint architecture plus the maintenance headache from maintaining two kinds of rules).

I'm happy to step back and let the TSC weigh in, but I can virtually guarantee this is a non-starter. ESLint depends on a full AST, period, there is no such thing as linting half a file.

**Comment by ilyavolodin** [comment]
AST is a tree, so running validation on "valid part" means running it on an incomplete branch of a tree. In some edge cases it might possibly work, but in most cases - it will not. This is an architectural decision of ESLint to only run rules on a valid AST, and I'm pretty sure that it's not something that will change.

**Comment by vitorbal** [comment]
Thanks for the suggestion, we really appreciate the interest in helping us improve ESLint. However, it seems like the team has a pretty clear position against this suggestion. I'm going to go ahead and close the issue, but feel free to discuss anything you feel is not clear yet.
